### ISSUE TITLE
zig cc -target x86_64-macos-none stopped working

### ISSUE THREAD

**Issue by markus-oberhumer** [ziglang/zig#19206]
### Zig Version

0.12.0-dev.3158+1e67f5021

### Steps to Reproduce and Observed Behavior

(I'm aware that there have been major macos code changes, so I waited several weeks before reporting. The issue has not been fixed yet, though.)

At https://github.com/upx/upx-test-build-with-zig we are continuously testing building UPX for 104 different build targets, and it seems that recently `zig cc -target x86_64-macos-none` stopped working.

Last version producing a **working** macos binary:

  - 0.12.0-dev.2341+92211135f
  - see `job-build-on-macos` log at https://github.com/upx/upx-test-build-with-zig/actions/runs/7662268830

All later Zig versions fail on macos, see logs at

 -  see `job-build-on-macos`  at https://github.com/upx/upx-test-build-with-zig/actions/workflows/build-upx-with-zig.yml

Note that binary artifacts are available for download for all our CI jobs.

### Expected Behavior

A working binary.

**Comment by kubkon** [comment]
New to me. Minimal repro/step-by-step instructions would be most appreciated!

**Comment by kubkon** [comment]
Thanks! I'll get on it right away so that it makes it into 0.12 release.

**Comment by kubkon** [comment]
I am trying to run the CI tests locally, but so far I cannot get passed missing artifacts:

```
$ make test
cd build/release && ctest
Test project /Users/kubkon/dev/upx/build/release
      Start  1: upx-version
 1/14 Test  #1: upx-version ......................   Passed    0.00 sec
      Start  2: upx-help
 2/14 Test  #2: upx-help .........................   Passed    0.00 sec
      Start  3: upx-sysinfo
 3/14 Test  #3: upx-sysinfo ......................   Passed    0.00 sec
      Start  4: upx-self-pack
 4/14 Test  #4: upx-self-pack ....................***Failed    0.00 sec
      Start  5: upx-self-pack-n2b
 5/14 Test  #5: upx-self-pack-n2b ................***Failed    0.00 sec
      Start  6: upx-self-pack-n2d
 6/14 Test  #6: upx-self-pack-n2d ................***Failed    0.00 sec
      Start  7: upx-self-pack-n2e
 7/14 Test  #7: upx-self-pack-n2e ................***Failed    0.00 sec
      Start  8: upx-self-pack-lzma
 8/14 Test  #8: upx-self-pack-lzma ...............***Failed    0.00 sec
      Start  9: upx-list
 9/14 Test  #9: upx-list .........................***Failed    0.00 sec
      Start 10: upx-fileinfo
10/14 Test #10: upx-fileinfo .....................***Failed    0.00 sec
      Start 11: upx-test
11/14 Test #11: upx-test .........................***Failed    0.00 sec
      Start 12: upx-unpack
12/14 Test #12: upx-unpack .......................***Failed    0.00 sec
      Start 13: upx-run-unpacked
Could not find executable ./upx-unpacked
Looked in the following places:
./upx-unpacked
./upx-unpacked
./Release/upx-unpacked
./Release/upx-unpacked
./Debug/upx-unpacked
./Debug/upx-unpacked
./MinSizeRel/upx-unpacked
./MinSizeRel/upx-unpacked
./RelWithDebInfo/upx-unpacked
./RelWithDebInfo/upx-unpacked
./Deployment/upx-unpacked
./Deployment/upx-unpacked
./Development/upx-unpacked
./Development/upx-unpacked
Unable to find executable: ./upx-unpacked
13/14 Test #13: upx-run-unpacked .................***Not Run   0.00 sec
      Start 14: upx-run-packed
Could not find executable ./upx-packed
Looked in the following places:
./upx-packed
./upx-packed
./Release/upx-packed
./Release/upx-packed
./Debug/upx-packed
./Debug/upx-packed
./MinSizeRel/upx-packed
./MinSizeRel/upx-packed
./RelWithDebInfo/upx-packed
./RelWithDebInfo/upx-packed
./Deployment/upx-packed
./Deployment/upx-packed
./Development/upx-packed
./Development/upx-packed
Unable to find executable: ./upx-packed
14/14 Test #14: upx-run-packed ...................***Not Run   0.00 sec

21% tests passed, 11 tests failed out of 14

Total Test time (real) =   0.03 sec

The following tests FAILED:
          4 - upx-self-pack (Failed)
          5 - upx-self-pack-n2b (Failed)
          6 - upx-self-pack-n2d (Failed)
          7 - upx-self-pack-n2e (Failed)
          8 - upx-self-pack-lzma (Failed)
          9 - upx-list (Failed)
         10 - upx-fileinfo (Failed)
         11 - upx-test (Failed)
         12 - upx-unpack (Failed)
         13 - upx-run-unpacked (Not Run)
         14 - upx-run-packed (Not Run)
Errors while running CTest
Output from these tests are in: /Users/kubkon/dev/upx/build/release/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.
make: *** [test] Error 8
```

Any help on getting the tests to actually execute would be appreciated.

**Comment by kubkon** [comment]
Hmm, so far building natively on x86_64 macos produces a working binary for me:

```
$ ./build/debug/upx --version
[doctest] doctest version is 2.4.11
[doctest] compiler version is clang 17.0.1 (64-bit, C++201703)
[doctest] run with "--dt-help" for options
===============================================================================
[doctest] test cases:  50 |  50 passed | 0 failed | 0 skipped
[doctest] assertions: 707 | 707 passed | 0 failed |
[doctest] Status: SUCCESS!
upx 4.2.2-devel.51+git-cfc5f42e
UCL data compression library 1.03
zlib data compression library 1.3.1.1-motley
LZMA SDK version 4.43
doctest C++ testing framework version 2.4.11
Copyright (C) 1996-2024 Markus Franz Xaver Johannes Oberhumer
Copyright (C) 1996-2024 Laszlo Molnar
Copyright (C) 2000-2024 John F. Reiser
Copyright (C) 2002-2024 Jens Medoch
Copyright (C) 1995-2024 Jean-loup Gailly and Mark Adler
Copyright (C) 1999-2006 Igor Pavlov
Copyright (C) 2016-2023 Viktor Kirilov
UPX comes with ABSOLUTELY NO WARRANTY; for details type 'upx -L'.
```

**Comment by kubkon** [comment]
Same with using `zld` as the linker with system tools:

```
$ CC="cc -fuse-ld=/Users/jakubkonka/dev/zld/zig-out/bin/ld64.zld -w" CXX="c++ -fuse-ld=/Users/jakubkonka/dev/zld/zig-out/bin/ld64.zld -w"  make build/debug
$ ./build/debug/upx --version
[doctest] doctest version is 2.4.11
[doctest] compiler version is clang 14.0.0 (64-bit, C++201703)
[doctest] run with "--dt-help" for options
===============================================================================
[doctest] test cases:  50 |  50 passed | 0 failed | 0 skipped
[doctest] assertions: 707 | 707 passed | 0 failed |
[doctest] Status: SUCCESS!
upx 4.2.2-devel.51+git-cfc5f42e
UCL data compression library 1.03
zlib data compression library 1.3.1.1-motley
LZMA SDK version 4.43
doctest C++ testing framework version 2.4.11
Copyright (C) 1996-2024 Markus Franz Xaver Johannes Oberhumer
Copyright (C) 1996-2024 Laszlo Molnar
Copyright (C) 2000-2024 John F. Reiser
Copyright (C) 2002-2024 Jens Medoch
Copyright (C) 1995-2024 Jean-loup Gailly and Mark Adler
Copyright (C) 1999-2006 Igor Pavlov
Copyright (C) 2016-2023 Viktor Kirilov
UPX comes with ABSOLUTELY NO WARRANTY; for details type 'upx -L'.
```

**Comment by kubkon** [comment]
In the former case, I've built Zig against 17.0.1 as seen in your doctest output. In the latter, I am overriding Apple's ld step with `ld64.zld`. In both cases, no segfault.

**Comment by kubkon** [comment]
Yes, I understand. I triple checked I was linking with Zig in the former case and zld in the latter case (I specifically set the tool version in MachO to 5 for Zig and 6 for zld to discriminate). Anyhow, I have followed the above instructions step-by-step, and using locally built Zig@`0.12.0-dev.3305+400145ee3`, no segfault:

<img width="869" alt="Screenshot 2024-03-14 at 19 54 39" src="https://github.com/ziglang/zig/assets/1519747/eb93728e-b4f7-400e-94a4-9406deaae9cb">

**Comment by kubkon** [comment]
What I will try next is building Zig using `zig-bootstrap` and see if that fails.

**Comment by kubkon** [comment]
That is incorrect. Depending on what patch version of LLVM you have at hand when building Zig from source, that's what gets embedded in the compiler.

**Comment by kubkon** [comment]
Indeed. Also, I have tested Apple's clang with standalone zld and this combination works fine. The code for both linkers, the one in Zig and the standalone one are virtually identical which starts suggesting this might not be a linker issue after all.

**Comment by kubkon** [comment]
Interestingly, `x86_64-macos` bootstrap contains 17.0.1 version of LLVM. Perhaps I really should test it on Linux targeting macOS.

**Comment by kubkon** [comment]
I see it affect all available macOS versions on GH.

**Comment by kubkon** [comment]
OK, I see that the CI's `zig-bootstrap` LLVM ranges from `17.0.0` to `17.0.1` while the prebuilt nightly reports `17.0.6` so clearly I am missing a step here. I will raise that with Andrew, but in the meantime I will build myself a copy of LLVM@17.0.6 locally and see if I can repro the issue.

**Comment by kubkon** [comment]
Sadly, no:

<img width="646" alt="Screenshot 2024-03-14 at 20 32 48" src="https://github.com/ziglang/zig/assets/1519747/aa111976-e277-4a01-8cc1-5208a3f6ef22">

**Comment by kubkon** [comment]
> Completely unrelated:
> 
> during my tests I discovered that zig does not work on macos-14 (it probably gets confused by some changes in Xcode-15.0.1)
> 
> Please see https://github.com/upx/upx-test-build-with-zig/actions/runs/8307582819/job/22736949690
> 
> and search for `error: unknown target CPU 'x86-64'`

Is that perhaps because macos-14 VM is arm64?

**Comment by kubkon** [comment]
> No, it is x86_64.
> 
> `uname -a` from the log says `Darwin Mac-1710586507180.local 22.6.0 Darwin Kernel Version 22.6.0: Sun Dec 17 22:18:09 PST 2023; root:xnu-8796.141.3.703.2~2/RELEASE_X86_64 x86_64`

That's interesting. I could have sworn GH is phasing out x86 macos support.

**Comment by kubkon** [comment]
> Please try for yourself.
> 
> In any case, the original issue seems fixed, so feel free to close as duplicate of #19221. Many thanks!

Taken from your macos-14 run:

![image](https://github.com/ziglang/zig/assets/1519747/64d0f1c3-555d-4110-8e65-872b0bffcab3)

**Comment by kubkon** [comment]
Duplicate of #19221 so closing as fixed.
